### PR TITLE
modify WEBPACKER PARALLEL environment variable.

### PR DESCRIPTION
**Pull request by masato-hi**
Will be possible to specify the WEBPACKER PARALLEL environment variable from the command line.

We also added a document about WEBPACKER_PARALLEL.

The master branch is currently working on 6.x, so we need to merge after merging this pull request.
https://github.com/rails/webpacker/pull/2839

I'm not good at English, so I'm looking for opinions on better documentation.

ref: https://github.com/rails/webpacker/pull/2093

### CODE REVIEW

File: package/utils/helpers.js [hunk 1]
```diff
@@ -37,11 +38,24 @@ const canProcess = (rule, fn) => {
   return null
 }
 
+const asBoolean = (str) => {
```
Comment (gauravtiwari) [inline package/utils/helpers.js:41]: thanks @masato-hi Could you please use the above function `isBoolean` (line 2)?

Comment (masato-hi) [reply package/utils/helpers.js:41]: Thanks for review.
Cannot pass `"false"` in the `isBoolean` function.

**Comment by pedrofurtado** [comment]
Hey, @masato-hi ! 👋 

Thanks for your contribution!

For this purpose (modify WEBPACKER_PARALLEL environment variable for a specific environment, for example), this works for your scenario?

- Edit your `config/webpack/[ENVIRONMENT].js` and put at the top the desired value of `WEBPACKER_PARALLEL`, like this snippet code below:
```javascript
process.env.WEBPACKER_PARALLEL = 'desired-value'
// ... existing code of file ...
```

Or this not works for you, and the PR in this case solves some problem related to it?

Let me know what you think about 🤝 🍻

**Comment by masato-hi** [comment]
@pedrofurtado Thanks for confirming.

It looks like it's working fine in my environment. :thinking:  (The `sed` command installed on MacOSX may not execute the following code successfully.)

```
rails new myapp --webpack
cd myapp
sed -i -e "s/gem 'webpacker', '~> 5.0'/gem 'webpacker', github: 'masato-hi\/webpacker', branch: 'modify-webpacker-parallel-env'/" Gemfile
bundle install
yes | bundle exec rake webpacker:install
sed -i -e 's/"@rails\/webpacker": ".*"/"@rails\/webpacker": "masato-hi\/webpacker#modify-webpacker-parallel-env"/' package.json
yarn
echo "
process.env.NODE_ENV = process.env.NODE_ENV || 'production'
process.env.WEBPACKER_PARALLEL = '2'

const webpackConfig = require('./base')

console.log(webpackConfig.optimization.minimizer[0])
module.exports = webpackConfig
" > config/webpack/production.js
rm .browserslistrc # cheap tricks for "Error [BrowserslistError]: contains both .browserslistrc and package.json with browsers"
RAILS_ENV=production bundle exec rake webpacker:clean webpacker:compile
```

Output:
```
Compiling...
Compiled all packs in /tmp/myapp/public/packs
TerserPlugin {
  options: {
    test: /\.[cm]?js(\?.*)?$/i,
    extractComments: true,
    parallel: 2,
    include: undefined,
    exclude: undefined,
    minify: undefined,
    terserOptions: {
      parse: [Object],
      compress: [Object],
      mangle: [Object],
      output: [Object]
    }
  }
}
asset manifest.json 23 bytes [emitted]
webpack 5.21.2 compiled successfully in 26 ms
```
The value of "parallel" is 2.

The configuration only affects production builds.
You also need to set process.env.WEBPACKER_PARALLEL before `require('@rails/webpacker')` / `require('./ base')` is executed.

**Comment by masato-hi** [comment]
@gauravtiwari 
The changes in docs/troubleshooting.md have been merged into master and are now reflected.
What do you think?

File: docs/troubleshooting.md [hunk 2]
```diff
@@ -111,6 +111,19 @@ mv $HOME/your_rails_app/node_modules/.bin/elm-make $HOME/your_rails_app/node_mod
...
+To solve this problem, you can specify the WEBPACKER_PARALLEL environment variable.
+Possible values ​​are true, false, or a number.
+
+```
+RAILS_ENV=production WEBPACKER_PARALLEL=false bundle exec rake webpacker:compile
+```
+
+Check the link for details.
+
+https://webpack.js.org/plugins/terser-webpack-plugin/#parallel
```
Comment (olleolleolle) [inline docs/troubleshooting.md:126]:
```suggestion
[`terser-webpack-plugin` documentation](https://webpack.js.org/plugins/terser-webpack-plugin/) has more details.
```

File: docs/troubleshooting.md [hunk 3]
```diff
@@ -111,6 +111,19 @@ mv $HOME/your_rails_app/node_modules/.bin/elm-make $HOME/your_rails_app/node_mod
 printf "#\041/bin/bash\n\necho \"Running elm-make with sysconfcpus -n 2\"\n\n$HOME/dependencies/sysconfcpus/bin/sysconfcpus -n 2 $HOME/your_rails_app/node_modules/.bin/elm-make-old \"\$@\"" > $HOME/your_rails_app/node_modules/.bin/elm-make
 chmod +x $HOME/your_rails_app/node_modules/.bin/elm-make
 ```
+## If an error occurs or times out during a production build on Continuous Integration (CI) services such as CircleCI, CodeShip, Travis CI
+
+The parallel option of Tersor-Plugin is enabled in the production build, which may cause memory shortage.
```
Comment (olleolleolle) [inline docs/troubleshooting.md:116]:
```suggestion
The parallel option of TerserWebpackPlugin is enabled in the production build, which may cause memory shortage.
```

File: docs/troubleshooting.md [hunk 4]
```diff
@@ -111,6 +111,19 @@ mv $HOME/your_rails_app/node_modules/.bin/elm-make $HOME/your_rails_app/node_mod
 printf "#\041/bin/bash\n\necho \"Running elm-make with sysconfcpus -n 2\"\n\n$HOME/dependencies/sysconfcpus/bin/sysconfcpus -n 2 $HOME/your_rails_app/node_modules/.bin/elm-make-old \"\$@\"" > $HOME/your_rails_app/node_modules/.bin/elm-make
 chmod +x $HOME/your_rails_app/node_modules/.bin/elm-make
 ```
+## If an error occurs or times out during a production build on Continuous Integration (CI) services such as CircleCI, CodeShip, Travis CI
```
Comment (olleolleolle) [inline docs/troubleshooting.md:114]:
```suggestion
## If an error occurs or times out during a production build
```

**Review by olleolleolle — COMMENTED** [review]
I made some inline suggestions, hope this helps.

**Comment by justin808** [comment]
@guillaumebriday, For sure, we should do something with PR. Running out of memory causes a very confusing build just seems to hang when it's swapping memory.

I don't like the current default: 
>        parallel: Number.parseInt(process.env.WEBPACKER_PARALLEL, 10) || true,

Defaults should be defaults, which is `true` for this option.

How about this:

1. default to true
2. all env override of process.env.WEBPACKER_PARALLEL to either be `false` or some lower number to set on CI machines?

https://webpack.js.org/plugins/terser-webpack-plugin/#parallel

> parallel
> Type: Boolean|Number Default: true
> 
> Use multi-process parallel running to improve the build speed. Default number of concurrent runs: os.cpus().length - 1.
> 
> ℹ️ Parallelization can speedup your build significantly and is therefore highly recommended.
> 
> ⚠️ If you use Circle CI or any other environment that doesn't provide real available count of CPUs then you need to setup explicitly number of CPUs to avoid Error: Call retries were exceeded (see #143, #202).

**Comment by justin808** [comment]
Feel free to reopen this PR here: 
https://github.com/shakacode/shakapacker/pulls

**Pull request closed without merging**